rollouts/pkg/packagediscovery/packagediscovery.go

```diff
@@ -0,0 +1,104 @@
+package packagediscovery
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path/filepath"
+
+	gitopsv1alpha1 "github.com/GoogleContainerTools/kpt/rollouts/api/v1alpha1"
+	"github.com/google/go-github/v48/github"
+	"golang.org/x/oauth2"
+	coreapi "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PackageDiscovery struct {
+	config    gitopsv1alpha1.PackagesConfig
+	client    client.Client
+	namespace string
+}
+
+type DiscoveredPackage struct {
+	Org       string
+	Name      string
+	Directory string
+	Revision  string
+}
+
+func NewPackageDiscovery(config gitopsv1alpha1.PackagesConfig, client client.Client, namespace string) *PackageDiscovery {
+	return &PackageDiscovery{
+		config:    config,
+		client:    client,
+		namespace: namespace,
+	}
+}
+
+func (d *PackageDiscovery) GetPackages(ctx context.Context) ([]DiscoveredPackage, error) {
+	gitRepoSelector := d.config.Git.GitRepoSelector
+	gitClient, err := d.getGitHubClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create git client: %w", err)
+	}
+
+	tree, _, err := gitClient.Git.GetTree(ctx, gitRepoSelector.Org, gitRepoSelector.Name, gitRepoSelector.Revision, true)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch tree from git: %w", err)
+	}
+
+	allPaths := []string{}
+	for _, entry := range tree.Entries {
+		if *entry.Type == "tree" {
+			allPaths = append(allPaths, *entry.Path)
+		}
+	}
+
+	packagesPaths := discoverPackagePaths(gitRepoSelector.PackagesPath, allPaths)
+
+	discoveredPackages := []DiscoveredPackage{}
+
+	for _, path := range packagesPaths {
+		thisDiscoveredPackage := DiscoveredPackage{Org: gitRepoSelector.Org, Name: gitRepoSelector.Name, Revision: gitRepoSelector.Revision, Directory: path}
+		discoveredPackages = append(discoveredPackages, thisDiscoveredPackage)
+	}
+
+	return discoveredPackages, nil
+}
+
+func (d *PackageDiscovery) getGitHubClient(ctx context.Context) (*github.Client, error) {
+	gitRepoSelector := d.config.Git.GitRepoSelector
+
+	httpClient := &http.Client{}
+
+	if secretName := gitRepoSelector.SecretRef.Name; secretName != "" {
```

Comment:
Thinking from API perspective, I think for `git` we can decouple the connectionInfo (auth in this case) from the different types of selectors. And we can do that as we evolve the code.